### DISCUSSION TITLE
Question on the perturb function

### DISCUSSION THREAD

**Discussion by yunjian-hash** [deepmodeling/dpdata#379] [Q&A]
Dear developers,

When the cell is perturbed, will the atomic coordinates change accordingly?  In addition, why perturb() function must require a positional argument: 'atom_pert_distance', if I only want the cell changes and the coordinates of atoms change accordingly, should I set the atom_pert_distance equals to 0?

Thanks!
![9059409b767e3103fda27557ae7e2d9](https://user-images.githubusercontent.com/75162924/204183106-2e94de51-8a81-44f7-8f8a-bbb43392c111.png)

**Comment by HuangJiameng** [comment]
Atoms will move about `atom_pert_distance` in random direction. And it can be set as 0. Source code is here, FYI.https://github.com/deepmodeling/dpdata/blob/2ac2f48b6761fb9b53236edcb533b3dc95ee264e/dpdata/system.py#L774-L826

**Comment by yunjian-hash** [comment]
> When the cell is perturbed, will the atomic coordinates change accordingly?

Thanks! How about the first questrion 'when the cell is perturbed, will the atomic coordinates change accordingly?',  the code does not tell about that.

**Comment by HuangJiameng** [comment]
I think that it does tell.
https://github.com/deepmodeling/dpdata/blob/2ac2f48b6761fb9b53236edcb533b3dc95ee264e/dpdata/system.py#L819-L820

**Reply by yunjian-hash** [reply]
Thanks! When I convert the system after pertubed, here comes an error. are the perturbed systems cannot be append to the multisystem?

![75f37deee854aed0e98d62a4eb5097f](https://user-images.githubusercontent.com/75162924/204192426-2977b638-1524-4ca7-9c51-20e14f228854.png)


from dpdata import System, LabeledSystem, MultiSystems
import numpy as np

####### 读入轨迹 #######

ls3=LabeledSystem('./Walker_2', fmt='cp2k/aimd_output')

####### 验证集结构数 #######

a3 = np.random.choice(len(ls3),size=100,replace=False)

####### 训练集结构数  #######

w3 = list(set(range(1,len(ls3)))-set(a3))[10:-1:6]
rs3=ls3.sub_system(w3)

####### 微扰训练集 #######
perturbed_system_1 = rs3.perturb(pert_num=1, cell_pert_fraction=0.1, atom_pert_distance=0, atom_pert_style='normal') 
perturbed_system_2 = rs3.perturb(pert_num=1, cell_pert_fraction=0.2, atom_pert_distance=0, atom_pert_style='normal') 
perturbed_system_3 = rs3.perturb(pert_num=1, cell_pert_fraction=0.3, atom_pert_distance=0, atom_pert_style='normal') 

####### 训练集 #######
ms= MultiSystems() 
ms.append(rs3) 
ms.append(perturbed_system_1)
ms.append(perturbed_system_2)
ms.append(perturbed_system_3)

ms.to_deepmd_raw('deepmd')
ms.to_deepmd_npy('deepmd',set_size=5000)

print('There are %d frames for training' % len(ls3.sub_system(w3)))
print('There are 3* %d frames for training with perturbed system' % len(perturbed_system_1))

**Reply by LiangWenshuo1118** [reply]
I guess this is because perturbed_systems only have coords and cells.  Energies, forces and virials information should also be contained.

**Reply by wanghan-iapcm** [reply]
referred to #381
